core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,320 @@
+      failuresForStage.foreach { case (exec, newFailures) =>
+        val prevFailures = executorIdToFailureCount.getOrElse(exec, 0)
+        val newTotal = prevFailures + newFailures.totalFailures
+
+        if (newTotal >= MAX_FAILURES_PER_EXEC) {
+          logInfo(s"Blacklisting executor $exec because it has $newTotal" +
+            s" task failures in successful task sets")
+          val now = clock.getTimeMillis()
+          val expiryTime = now + EXECUTOR_RECOVERY_MILLIS
+          executorIdToBlacklistExpiryTime.put(exec, expiryTime)
+          executorIdToFailureCount.remove(exec)
+          if (expiryTime < nextExpiryTime) {
+            nextExpiryTime = expiryTime
+          }
+
+          val node = scheduler.getHostForExecutor(exec)
+          val execs = scheduler.getExecutorsAliveOnHost(node).getOrElse(Set())
+          val blacklistedExecs = execs.filter(executorIdToBlacklistExpiryTime.contains(_))
+          if (blacklistedExecs.size >= MAX_FAILED_EXEC_PER_NODE) {
+            logInfo(s"Blacklisting node $node because it has ${blacklistedExecs.size} executors " +
+              s"blacklisted: ${blacklistedExecs}")
+            nodeIdToBlacklistExpiryTime.put(node, expiryTime)
+            // make a copy of the blacklisted nodes so nodeBlacklist() is threadsafe
+            _nodeBlacklist.set(nodeIdToBlacklistExpiryTime.keySet.toSet)
+          }
+        } else {
+          executorIdToFailureCount.put(exec, newTotal)
+        }
+      }
+    }
+    // when we blacklist a node within a stage, we don't directly promote that node to being
+    // blacklisted for the app.  Instead, we use the mechanism above to decide whether or not to
+    // blacklist any executors for the app, and when doing so we'll check whether or not to also
+    // blacklist the node.  That is why we just remove this entry without doing any promotion to
+    // the full app blacklist.
+    stageIdToBlacklistedNodes.remove(stageId)
+  }
+
+  def taskSetFailed(stageId: Int): Unit = {
+    // just throw away all the info for the failures in this taskSet -- assume the executors were
+    // fine, the failures were just b/c the taskSet itself was bad (eg., bad user code)
+    stageIdToExecToFailures.remove(stageId)
+    stageIdToBlacklistedNodes.remove(stageId)
+  }
+
+  /**
+   * Return true if this executor is blacklisted for the given stage.  Completely ignores whether
+   * the executor is blacklisted overall (or anything to do with the node the executor is on).
+   */
+  def isExecutorBlacklistedForStage(
+      stageId: Int,
+      executorId: String): Boolean = {
+    stageIdToExecToFailures.get(stageId).flatMap(_.get(executorId))
+      .map(_.totalFailures >= MAX_FAILURES_PER_EXEC_STAGE).getOrElse(false)
+  }
+
+  def isExecutorBlacklisted(executorId: String): Boolean = {
+    executorIdToBlacklistExpiryTime.contains(executorId)
+  }
+
+  def isNodeBlacklistedForStage(node: String, stageId: Int): Boolean = {
```

Comment:
Any special reason why this (and `isExecutorBlacklistedForStage`) has the `ForStage` suffix while the other almost-overloaded versions don't?